lark/exceptions.py

```diff
@@ -109,32 +110,54 @@ def match_examples(self, parse_fn, examples, token_type_match_fallback=False, us
                             candidate = label, False
 
         return candidate[0]
+    
+    def _format_terminals(self, names):
+        if self._all_terminals:
+            t = []
+            for name in names:
+                try:
+                    t.append(next(t.nice_print for t in self._all_terminals if t.name == name))
+                except StopIteration:
+                    # If we don't find the corresponding Terminal (which *should* never happen), don't error.
+                    # Broken __str__ for Exception are some of the worst bugs
+                    t.append(t.display_name)
+        else:
+            t = names
+        return "Expected one of: \n\t* %s\n" % '\n\t* '.join(t)
+
 
 
 class UnexpectedCharacters(LexError, UnexpectedInput):
-    def __init__(self, seq, lex_pos, line, column, allowed=None, considered_tokens=None, state=None, token_history=None):
+    def __init__(self, seq, lex_pos, line, column, allowed=None, considered_tokens=None, state=None, token_history=None, _all_terminals=None):
         self.line = line
         self.column = column
         self.pos_in_stream = lex_pos
         self.state = state
+        self._all_terminals = _all_terminals
 
         self.allowed = allowed
         self.considered_tokens = considered_tokens
+        self.token_history = token_history
 
         if isinstance(seq, bytes):
-            _s = seq[lex_pos:lex_pos+1].decode("ascii", "backslashreplace")
+            self._s = seq[lex_pos:lex_pos+1].decode("ascii", "backslashreplace")
         else:
-            _s = seq[lex_pos]
-
-        message = "No terminal defined for '%s' at line %d col %d" % (_s, line, column)
-        message += '\n\n' + self.get_context(seq)
-        if allowed:
-            message += '\nExpecting: %s\n' % allowed
-        if token_history:
-            message += '\nPrevious tokens: %s\n' % ', '.join(repr(t) for t in token_history)
-
-        super(UnexpectedCharacters, self).__init__(message)
-
+            self._s = seq[lex_pos]
+        self._context = self.get_context(seq)
+        
+        super(UnexpectedCharacters, self).__init__()
+
+    def __str__(self):
```

Comment:
It's probably better to have a `_str()` method, and have `__str__(self): return self._str()`. That way you can debug it with less pain.

Reply:
I don't really see a benefit of `e._str()` over `str(e)`.